tools/src/main/java/org/apache/kafka/trogdor/workload/RoundTripWorker.java

```diff
@@ -0,0 +1,328 @@
+    }
+
+    class ValidateSpec implements Runnable {
+        @Override
+        public void run() {
+            try {
+                if (spec.targetMessagesPerSec() <= 0) {
+                    throw new ConfigException("Can't have targetMessagesPerSec <= 0.");
+                }
+                executor.submit(new ProducerRunnable());
+                executor.submit(new ConsumerRunnable());
+            } catch (Exception e) {
+                WorkerUtils.abort("ValidateSpec", e, doneFuture);
+            }
+        }
+    }
+
+    private static class ToSendTrackerResult {
+        final int index;
+        final boolean firstSend;
+
+        ToSendTrackerResult(int index, boolean firstSend) {
+            this.index = index;
+            this.firstSend = firstSend;
+        }
+    }
+
+    private static class ToSendTracker {
+        private final int maxMessages;
+        private final List<Integer> failed = new ArrayList<>();
+        private int frontier = 0;
+
+        ToSendTracker(int maxMessages) {
+            this.maxMessages = maxMessages;
+        }
+
+        synchronized void addFailed(int index) {
+            failed.add(index);
+        }
+
+        synchronized ToSendTrackerResult next() {
+            if (failed.isEmpty()) {
+                if (frontier >= maxMessages) {
+                    return null;
+                } else {
+                    return new ToSendTrackerResult(frontier++, true);
+                }
+            } else {
+                return new ToSendTrackerResult(failed.remove(0), false);
+            }
+        }
+    }
+
+    class ProducerRunnable implements Runnable {
+        private final Throttle throttle;
+
+        ProducerRunnable() {
+            Properties props = new Properties();
+            props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, spec.bootstrapServers());
+            props.put(ProducerConfig.BATCH_SIZE_CONFIG, 16 * 1024);
+            props.put(ProducerConfig.BUFFER_MEMORY_CONFIG, 4 * 16 * 1024L);
+            props.put(ProducerConfig.MAX_BLOCK_MS_CONFIG, 1000L);
+            props.put(ProducerConfig.CLIENT_ID_CONFIG, "producer." + id);
+            props.put(ProducerConfig.ACKS_CONFIG, "all");
+            props.put(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG, 105000);
+            producer = new KafkaProducer<>(props, new StringSerializer(),
+                new ByteArraySerializer());
+            int perPeriod = WorkerUtils.
+                perSecToPerPeriod(spec.targetMessagesPerSec(), THROTTLE_PERIOD_MS);
+            this.throttle = new Throttle(perPeriod, THROTTLE_PERIOD_MS);
+        }
+
+        @Override
+        public void run() {
+            byte[] value = new byte[VALUE_SIZE];
+            final ToSendTracker toSendTracker = new ToSendTracker(spec.maxMessages());
+            long messagesSent = 0;
+            long uniqueMessagesSent = 0;
+            log.debug("{}: Starting RoundTripWorker#ProducerRunnable.", id);
```

Comment:
Is there a reason why you're not using `LogContext` instead of manually adding the `id` in every log message?

Reply:
Trogdor hasn't been converted to using LogContexts yet.  It does seem like it would be a good idea, though.  Is it worth doing in a separate PR?

Reply:
Sounds good.